src/OrbitGl/ThreadTrack.cpp

```diff
@@ -405,3 +397,100 @@ float ThreadTrack::GetHeaderHeight() const {
   header_height += std::max(0, track_count - 1) * space_between_subtracks;
   return header_height;
 }
+
+void ThreadTrack::OnTimer(const TimerInfo& timer_info) {
+  TimerTrack::OnTimer(timer_info);
+
+  // Get the address of the freshly added TextBox and insert it into the ScopeTree.
+  std::shared_ptr<TimerChain> timer_chain = timers_[timer_info.depth()];
+  CHECK(timer_chain != nullptr);
+  {
+    absl::MutexLock lock(&scope_tree_mutex_);
+    scope_tree_.Insert(timer_chain->Last());
+  }
+}
+
+[[nodiscard]] static inline internal::Rect GetTimerInfoRect(const TimerInfo& timer_info,
+                                                            const internal::DrawData& draw_data,
+                                                            const TimeGraph* time_graph,
+                                                            float world_pos_y, float world_size_y) {
+  double start_us = time_graph->GetUsFromTick(timer_info.start());
+  double end_us = time_graph->GetUsFromTick(timer_info.end());
+  double elapsed_us = end_us - start_us;
+  double normalized_start = start_us * draw_data.inv_time_window;
+  double normalized_length = elapsed_us * draw_data.inv_time_window;
+  float world_timer_width = static_cast<float>(normalized_length * draw_data.world_width);
+  float world_timer_x =
+      static_cast<float>(draw_data.world_start_x + normalized_start * draw_data.world_width);
+  return internal::Rect(world_timer_x, world_pos_y, world_timer_width, world_size_y);
+}
+
+[[nodiscard]] static inline uint64_t GetNextPixelBoundaryTimeNs(
+    float world_x, const internal::DrawData& draw_data) {
+  float normalized_x = (world_x - draw_data.world_start_x) / draw_data.world_width;
+  int pixel_x = static_cast<int>(ceil(normalized_x * draw_data.canvas->GetWidth()));
+  return draw_data.min_tick + pixel_x * draw_data.ns_per_pixel;
+}
+
+// We minimize overdraw when drawing lines for small events by discarding events that would just
+// draw over an already drawn pixel line. When zoomed in enough that all events are drawn as boxes,
+// this has no effect. When zoomed  out, many events will be discarded quickly.
+void ThreadTrack::UpdatePrimitives(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                   PickingMode picking_mode, float z_offset) {
+  UpdatePrimitivesOfSubtracks(batcher, min_tick, max_tick, picking_mode, z_offset);
+  UpdateBoxHeight();
+  visible_timer_count_ = 0;
+  internal::DrawData draw_data = GetDrawData(
```

Comment:
Again, const?

Reply:
Here I find it helps understand that the draw_data is read-only in a relatively big function where draw_data is used at multiple spots. I added const, thanks.